Prepare to relax the constraints of PodDNSConfig

gjkim42 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug
/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

This PR 
- prepares to relax the constraints of PodDNSConfig
- relaxes the constraints of resolv.conf in kubelet
- additionally, validates the length of the DNS search path in kubelet

More details,
this PR affects 
1. api-server validating already created Pod's DNSConfig to tolerate the Pod from the next release
1.  kubelet validating [`resolvConf`](https://github.com/kubernetes/kubernetes/blob/bb89384f3981bae3f1d507c340593514abc734bc/staging/src/k8s.io/kubelet/config/v1beta1/types.go#L516-L523) to allow expanded DNS config
1. kubelet to allow expanded DNS config when validating actual DNS resolver configuration composed of 
    - cluster domain suffixes(e.g. default.svc.cluster.local, svc.cluster.local, cluster.local)
    - kubelet's `resolvConf`
    - Pod's DNSConfig

Alternative: #100651 

Background: The limit of the number of DNS search paths and length of the DNS search list has been removed from libc.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubernetes/issues/100582
Tracking issue: https://github.com/kubernetes/kubernetes/issues/100622

#### Special notes for your reviewer:

https://access.redhat.com/solutions/58028

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
kubelet: Allow expanded DNS configuration
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

gjkim42 commented:
/cc @liggitt @dchen1107 @soltysh 
Could you take a look at this when available?

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3001,10 +2999,6 @@ func validatePodDNSConfig(dnsConfig *core.PodDNSConfig, dnsPolicy *core.DNSPolic
 			}
 		}
-		// Validate searches.
-		if len(dnsConfig.Searches) > MaxDNSSearchPaths {
```
I'll defer approval of the runtime implications of this PR to node/network folks.

for the API implications, if we relax validation, it has to be done over two releases so that we don't allow data to be persisted the previous version of the API server considers invalid. 

Release 1: tolerate objects with more search paths on update, but keep the current limit when creating objects or updating objects that meet the current limit

release 2: relax the limit unconditionally

liggitt replied:
Also, do we want to remove the limit completely and make an unbounded list (unlikely), or relax it to be larger?

gjkim42 replied:
@liggitt 
Thanks for your comment.

> it has to be done over two releases

Did you mean that this patch has to be done over two **_minor_** releases, right? 

> Release 1: tolerate objects with more search paths on update, but keep the current limit when creating objects or updating objects that meet the current limit

I am not sure about this. Are you saying that objects should be created with the current limit but they can be updated beyond the limit?

gjkim42 replied:
I want to enable it in 1.21.x as an alpha gate. If we decide to proceed as an alpha gate, do we need a KEP process?

aojea replied:
As Tim said, is too late for 1.21 and you can not merge into a minor release new features https://github.com/kubernetes/kubernetes/pull/100583#issuecomment-808589281 , it has to go in 1.22.

Agree with Tim that a small KEP with a feature gate seems better and cleaner than adding the `isLegacy` argument to the validation functions, also, this is similar to the `kubelete exec probes issue`, and they implemented the KEP + FeatureGate solution  too https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1972-kubelet-exec-probe-timeouts

gjkim42 commented:
/cc @ehashman @SergeyKanzhelev (sig-node)
Could you PTAL? Thanks.

thockin commented on pkg/kubelet/network/dns/dns.go:
```diff
@@ -173,19 +168,6 @@ func (c *Configurer) CheckLimitsForResolvConf() {
-		return
-	}
-
 	if len(strings.Join(hostSearch, " ")) > validation.MaxDNSSearchListChars {
```
I think you can kill this off too, but it would be nice to confirm.

thockin commented:
It's too late for 1.21, unfortunately.

On Fri, Mar 26, 2021 at 4:17 PM Geon-Ju Kim ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In pkg/apis/core/validation/validation.go
> <https://github.com/kubernetes/kubernetes/pull/100583#discussion_r602628151>
> :
>
> > @@ -3001,10 +2999,6 @@ func validatePodDNSConfig(dnsConfig *core.PodDNSConfig, dnsPolicy *core.DNSPolic
>  				allErrs = append(allErrs, field.Invalid(fldPath.Child("nameservers").Index(i), ns, "must be valid IP address"))
>  			}
>  		}
> -		// Validate searches.
> -		if len(dnsConfig.Searches) > MaxDNSSearchPaths {
>
> I want to enable it in 1.21.x as an alpha gate. If we decide to proceed as
> an alpha gate, do we need a KEP process?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/100583#discussion_r602628151>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVDP4CNPKHA5QKIOJITTFUI2JANCNFSM4Z3HC2XQ>
> .
>

BenTheElder commented:
What about resolvers other than glibc?

thockin commented:
I did some playing and nobody seems to explode.  They just ignore the
overage.  I didn't not claim to have been exhaustive but I have to assume
they don't just abort() on bad input. :)

Am I too optimistic?

On Fri, Mar 26, 2021, 10:10 PM Benjamin Elder ***@***.***>
wrote:

> What about resolvers other than glibc?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/100583#issuecomment-808656011>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVD4JE5QYXMOHM3CYTTTFVSDBANCNFSM4Z3HC2XQ>
> .
>

gjkim42 commented:
I tested some scenarios at the older version.
(CentOS 6.10, glibc 2.12)

My observation is similar to @thockin 's
```bash
$ cat /etc/resolv.conf
cat /etc/resolv.conf 
search 111 222 333 444 555 google.com redhat.com
options timeout:1 attempts:2
nameserver 8.8.8.8

$ ping -c 1 drive.google.com
PING drive.google.com (172.217.25.78) 56(84) bytes of data.
64 bytes from nrt13s50-in-f14.1e100.net (172.217.25.78): icmp_seq=1 ttl=113 time=85.0 ms

--- drive.google.com ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 85.086/85.086/85.086/0.000 ms

$ ping -c 1 people.redhat.com
PING people.redhat.com (209.132.183.19) 56(84) bytes of data.
64 bytes from people.redhat.com (209.132.183.19): icmp_seq=1 ttl=230 time=147 ms

--- people.redhat.com ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 147.081/147.081/147.081/0.000 ms

$ ping -c 1 drive
PING drive.google.com (172.217.25.78) 56(84) bytes of data.
64 bytes from nrt13s50-in-f14.1e100.net (172.217.25.78): icmp_seq=1 ttl=113 time=39.5 ms

--- drive.google.com ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 39.599/39.599/39.599/0.000 ms

$ ping -c 1 people
ping: unknown host people
```
---

@BenTheElder 
I am not familiar with the mechanics of a DNS resolving.
Is there anything else that I should test?

BenTheElder commented:
I would be curious about seeing at least:
- glibc
- musl
- golang without libc

I'm not sure if there are any other stacks in widespread use in linux contsiners but those three definitely are.

gjkim42 commented:
@BenTheElder 
Thanks for your comment.
Now I understand what you said.

I've tried some tests at
- `old glibc(centos 7.0.1406)`: ignore the outrage
- `glibc(centos 7.9)`: allow more than 7 search paths
- `old musl(alpine 2.6)`: resolvconf not works ??
- `musl(alpine 3)`: allow more than 7 search paths
- `pure Go resolver(go 1.10 and go 1.16)`: allow more than 7 search paths

I would say that at least they don't have any critical issue.
If we worry about this, we could proceed as a feature gate.

Need your advice.

gjkim42 commented:
@thockin 
According to MaxDNSSearchListChars, we can't relax MaxDNSSearchListChars.
Only pure go DNS resolver allows search lists of more than 256 characters.

- glibc 2.17-323(centos 7.9) ignores the outrage.
- musl 1.2.2(alpine 7.9) ignores the outrage.
- Debian glibc 2.28-10(Debian 10): the `ping` gets **SIGABRT** if search lists are more than 256 characters.
```
root@golang:/go# cat /etc/resolv.conf 
search google.com aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa google.com redhat.com
options ndots:5
nameserver 8.8.8.8

root@golang:/go# ping drive
ping: resolv_conf.c:571: update_from_conf: Assertion `resolv_conf_matches (resp, conf)' failed.
Aborted (core dumped)

root@golang:/go# nslookup drive
Server:         8.8.8.8
Address:        8.8.8.8#53

Non-authoritative answer:
Name:   drive.google.com
Address: 172.217.25.206
Name:   drive.google.com
Address: 2404:6800:4004:81a::200e

root@golang:/go# nslookup people
nslookup: 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' is not a legal IDN name (domain label longer than 63 characters), use +noidnin
```

gjkim42 commented:
Hid the previous comment. (It is WRONG)

Regarding MaxDNSSearchListChars, 
I confirmed that, 
- `glibc 2.17-55`: ignores overage(255 characters)
- `glibc 2.17-323`: allows a search path list above 255
- `glibc 2.28`: 
    - allows a search path list above 255
    - `ping` gets **SIGABRT** if there is **A SEARCH PATH** above 255
- `musl 1.2.2`: ignores the overage
- `pure Go resolver(go 1.10 and go 1.16)`: allows a search path list above 255

gjkim42 committed:
Prepare to relax the constraints of PodDNSConfig

This commit also relaxes the constraints of resolv.conf in kubelet.

The limit of the number of DNS search paths and length of the
DNS search list has been removed from libc.

gjkim42 committed:
kubelet: Validate the length of the DNS search path

gjkim42 commented:
/check-cla

gjkim42 commented:
/retest

gjkim42 commented:
/retest

gjkim42 commented:
Done.
Updated the PR description.

```
This PR
- prepares to relax the constraints of PodDNSConfig
- relaxes the constraints of resolv.conf in kubelet
- additionally, validates the length of the DNS search path in kubelet
```

/assign @thockin 
/cc @dims @mrunalp @liggitt 
Could you PTAL? Thanks.

dims commented:
/assign @aojea

liggitt commented on pkg/apis/apps/validation/validation.go:
```diff
@@ -303,7 +303,7 @@ func ValidateDaemonSetStatusUpdate(ds, oldDS *apps.DaemonSet) field.ErrorList {
 
 // ValidateDaemonSetSpec tests if required fields in the DaemonSetSpec are set.
-func ValidateDaemonSetSpec(spec *apps.DaemonSetSpec, fldPath *field.Path, opts apivalidation.PodValidationOptions) field.ErrorList {
+func ValidateDaemonSetSpec(spec *apps.DaemonSetSpec, fldPath *field.Path, opts apivalidation.PodValidationOptions, isLegacy bool /* TODO: after 1.22, remove isLegacy */) field.ErrorList {
```
make this a field in PodValidationOptions that defaults safe (e.g. `AllowLongerDNSSearchPaths`), not a separate boolean field we have to plumb everywhere

populate the field in GetValidationOptionsFromPodSpecAndMeta based on the presence of more DNS search paths or longer search paths in the existing pod spec

gjkim42 replied:
I will fix them all. Thanks!

gjkim42 replied:
Done.

@aojea 
Cleaned up the code.
How about now?

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2962,10 +2962,16 @@ const (
-	// Max number of characters in search path.
-	MaxDNSSearchListChars = 256
+	// Max number of domains in the search path list.
+	MaxDNSSearchPaths = 32
```
suggest picking new names (MaxDNSSearchPathsExpanded, MaxDNSSearchPathsLegacy) rather than changing constant values in ways that can silently break existing consumers

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2978,7 +2984,7 @@ func validateReadinessGates(readinessGates []core.PodReadinessGate, fldPath *fie
 }
 
-func validatePodDNSConfig(dnsConfig *core.PodDNSConfig, dnsPolicy *core.DNSPolicy, fldPath *field.Path) field.ErrorList {
+func validatePodDNSConfig(dnsConfig *core.PodDNSConfig, dnsPolicy *core.DNSPolicy, fldPath *field.Path, allowLiberalDNSConfig bool /* TODO: after 1.22, remove this */) field.ErrorList {
```
plumb PodValidationOptions here instead of just the boolean

gjkim42 replied:
Do we need to pass the whole PodValidationOptions?

liggitt commented on pkg/kubelet/network/dns/dns.go:
```diff
@@ -99,6 +100,14 @@ func omitDuplicates(strs []string) []string {
 func (c *Configurer) formDNSSearchFitsLimits(composedSearch []string, pod *v1.Pod) []string {
 	limitsExceeded := false
 
+	for i, search := range composedSearch {
```
does this mean 1.21 and 1.22 kubelets can break when running a pod with expanded DNS config allowed by a 1.23 API server?

gjkim42 replied:
Not exactly. 
A pod's search path will be validated by the API server to not exceed 253 characters, but kubelet's resolv.conf can contain a search path of more than 253 characters at 1.22 kubelet. 1.22 kubelet has to validate this.

gjkim42 commented:
@liggitt @thockin
Done.
Could you PTAL?

gjkim42 commented:
/cc @deads2k @ehashman 
Could you PTAL? Thanks

aojea commented:
@gjkim42 do we have guarantee that the `cluster domain suffixes`  are going to be always in the first 6 search domains? I think that if that is not guarantee, and the cluster domains are out of those 6 first domains,  pods ons Kubernetes clusters with old glibc versions will not be able to resolve ... 

Another naive question is, how are Windows clusters affected? I really don't have any idea how Windows systems work with resolv.conf.

I really think that this change deserves a KEP and a feature gate to opt-in for this new behaviour, so we can enable this once we have clear that this doesn't explode, we can't evaluate all distros, developer languages, ...  I can't imagine how to solve the problem if we enable this feature directly and after some months we found that something breaks :shrug:

gjkim42 commented:
> do we have guarantee that the `cluster domain suffixes` are going to be always in the first 6 search domains?

If there are, they will be the first 3 search domains.



> Another naive question is, how are Windows clusters affected? I really don't have any idea how Windows systems work with resolv.conf.

I have no idea too. We should inspect this.



> I really think that this change deserves a KEP and a feature gate to opt-in for this new behaviour, so we can enable this once we have clear that this doesn't explode, we can't evaluate all distros, developer languages, ... I can't imagine how to solve the problem if we enable this feature directly and after some months we found that something breaks 🤷

Agree. I will make an issue in the KEP repo. Thanks for your comment.

thockin commented:
I think #100651 is the better approach

gjkim42 commented:
Closed in favor of #100651
